infoschema/infoschema.go

```diff
@@ -405,16 +407,32 @@ func GetInfoSchemaBySessionVars(sessVar *variable.SessionVars) InfoSchema {
 }
 
 func (is *infoSchema) BundleByName(name string) (*placement.Bundle, bool) {
+	is.ruleBundleMutex.RLock()
+	defer is.ruleBundleMutex.RUnlock()
 	t, r := is.ruleBundleMap[name]
 	return t, r
 }
 
-func (is *infoSchema) RuleBundles() map[string]*placement.Bundle {
-	return is.ruleBundleMap
+func (is *infoSchema) RuleBundles() []*placement.Bundle {
+	is.ruleBundleMutex.RLock()
+	defer is.ruleBundleMutex.RUnlock()
+	bundles := make([]*placement.Bundle, 0, len(is.ruleBundleMap))
+	for _, bundle := range is.ruleBundleMap {
+		bundles = append(bundles, bundle)
+	}
+	return bundles
+}
+
+func (is *infoSchema) SetBundle(bundle *placement.Bundle) {
+	is.ruleBundleMutex.Lock()
+	defer is.ruleBundleMutex.Unlock()
+	is.ruleBundleMap[bundle.ID] = bundle
```

Comment:
How do you guarantee that the map is initialized?

Reply:
I did not promise, I assume, and there is only one initialization from `infoschema/builder.go`. So it should be initialized unless you are not using the infoschema builder to build infoschema

```
infoschema/builder.go:
ruleBundleMap:       map[string]*placement.Bundle{},
```

Reply:
Please check and initialize it in the interfaces. Do not assume the callers of these interfaces.

Reply:
OK. But I do think there is a need for a standard rule or common sense among different modules/coders: just make sure everyone will initialize the struct by its new methods. Otherwise, there are redundant checks everywhere.

Reply:
The thing is, some creation of `infoSchema` doesn't initialize `ruleBundleMap`, such as `MockInfoSchema` and `MockInfoSchemaWithSchemaVer`. It's hard to say the structs created by those functions won't be used to get or set rule bundles in the future.

Reply:
Then, these methods should use a consistent way to initialize `infoSchema`. IMO, this is just another assumption/common sense: `infoSchema` is not initialized by some unified/standard methods.

But simplicity favors regularity. I would choose to standardize the initialization of `infoSchema`, which will lead to an always non-nil/valid state of `ruleBundleMap`. Possibly, some refactoring on the `infoschema builder` API would work.

I could file an issue and look into this if you think it is a good idea.